### DISCUSSION TITLE
Change the order of the completion, e.g. based on the extension or size of the entry

### DISCUSSION THREAD

**Discussion by tobiasBora** [marlonrichert/zsh-autocomplete#642] [Q&A]
## What do you want?

Would it be possible to allow the user to change the order of the suggested completion, or even to move some item in a "less important" section? E.g. in:

![image](https://github.com/marlonrichert/zsh-autocomplete/assets/2164118/de734a72-a4a1-443f-9337-1664bf6ec572)

I’d prefer to have:
1. all `.tex` files appear first, before the .pdf (these are the "important extensions" I care about)
2. the .bbl, .run.xml, .aux, .synctex.gz… appear in another section "less important files" rather than "files" since I never directly manipulate them and they pollute my view
3. The "less important files" should also be determined by regexp, so that I can add for instance all files that contain "-pratend".
4. if two files share the same base name like `hey.tex` and `hey-backup.tex` or `hey-pratend-foo.tex`, I would prefer to have the shorter files listed first: sometimes, I can even type the full name, and the file is listed at the very end of the line, making it hard to select:

![image](https://github.com/marlonrichert/zsh-autocomplete/assets/2164118/cfa14d2d-f154-4724-86da-5037002a1e7e)

(not that these screenshots are taken with a relatively old version (~1y) so things might have changed a bit now, but I see nothing in the doc about that anyway)

## Why do you want this?

In LaTeX, you get an enormous amount of temporary files created during compilation, and most of the time only the `.tex` file and the `.pdf` files are important. Sometimes, some `.tex` files are also created with the same prefix but with an additional part in the name (the `-pratend` in the screenshot). This means that completion of these kind of files is very hard, and it is often easier to type the full name manually.

## Who else would benefit from this?

People that care about some extensions more than others, so at least any LaTeX user.

## How should it work?

The user should provide in configuration something like:
```
# first method (advanced users)
zsh_sort_function() { write the function that sets the category of each file and their position (for advanced users only)}
# second method (regular users)
- importantFiles: [ "\.tex$", "\.pdf$" ] # this will list tex before pdf
- lessImportantFiles: [ "\.bbl$", "\.synctex$", "-pratend-" ]
- priorityShortFiles: true
```

When I perform the following steps:
1. Type a string with proposed completion

Then I expect the following to happen:
* files that match the list "importantFiles" should be listed before other files (unless maybe if the fuzzy matching is way more fuzzy for the normal files?)
* files that match the list "lessImportantFiles" should appear in a separate section "Less important files", below the normal "file" section.
* shorter files should appear first

**Comment by marlonrichert** [comment]
Those things are already possible.

> 1. all `.tex` files appear first, before the .pdf (these are the "important extensions" I care about)
> 2. the .bbl, .run.xml, .aux, .synctex.gz… appear in another section "less important files" rather than "files" since I never directly manipulate them and they pollute my view
> 3. The "less important files" should also be determined by regexp, so that I can add for instance all files that contain "-pratend".

You can tell Zsh's completion system to hide files matching a certain pattern, unless you type something that matches only those files:

```sh
zstyle ':completion:*files' ignored-patterns '*.'{aux,bbl,pdf,run.xml,synctext.gz} '*-pratend'
```

Read [`zsh compsys` > Completion System Configuration](https://zsh.sourceforge.io/Doc/Release/Completion-System.html#index-ignored_002dpatterns_002c-completion-style) for more info.

> 4. if two files share the same base name like `hey.tex` and `hey-backup.tex` or `hey-pratend-foo.tex`, I would prefer to have the shorter files listed first: sometimes, I can even type the full name, and the file is listed at the very end of the line, making it hard to select:

That depends on the value of `LC_COLLATE` or, if not set, `LC_ALL` or, if not set, `LANG`. What the value does exactly is dependent on your OS.

I find that this gives good sorting results on Linux

```sh
export LC_COLLATE=C.UTF-8
```

whereas on macOS, I get better results with

```sh
export LANG=en_US.UTF-8
```

Read IBM's article _[Understanding locale environment variables](https://www.ibm.com/docs/en/aix/7.3?topic=locales-understanding-locale-environment-variables)_ for more info.

**Reply by tobiasBora** [reply]
Thanks for your answer. Thank you for the link to ignored-patterns, it is not exactly what I wanted (e.g. I can't say "list tex before pdf", or move the items to another section to still see them) but it is a first start. But unfortunately I cannot make it work, have I done something stupid?

![image](https://github.com/marlonrichert/zsh-autocomplete/assets/2164118/8ba06abc-ff30-41dd-b2e1-407fd1535ff9)

and same for `export LC_COLLATE=C.UTF-8`, it does not change anything, and I guess it will not be much customizable…

**Reply by marlonrichert** [reply]
Sorry, `':completion:*:files'` should've been `':completion:*files'`. I fixed it now in my answer.
